### ISSUE TITLE
Puma cluster high CPU usage

### ISSUE THREAD

**Issue by mwalsher** [puma/puma#755]
Having an issue with puma (cluster mode with 3 workers, 1-8 threads) where two workers are each utilizing 100% CPU in what seems to be a deadlock. We're using version 2.6.0, ruby 2.0.0, and rails 4.0.0).

`strace` shows the following output in repeating fashion:

```
sched_yield()                           = 0
futex(0x102ff64, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x102ff60, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x102ff30, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x102ff98, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x102ff9c, FUTEX_WAIT_PRIVATE, 94471, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x102ff30, FUTEX_WAKE_PRIVATE, 1) = 0
sched_yield()                           = 0
```

Restarting the app fixes the problem for a while, but whenever the server gets busy puma seems to randomly get hung.

puma.rb config is: https://gist.github.com/mwalsher/8adf101b8492bec65a6d

Any ideas?

**Comment by camallen** [comment]
Yeah I'm seeing something very similar, 1 worker default threads and 1 thread gets stuck with high cpu that slows all other threads on that node. It's been happening for a while on our api nodes and a restart of puma fixes it but they come back. 

Have you tried debugging the process / threads using gdb to get backtraces? [This article](https://blog.newrelic.com/2013/04/29/debugging-stuck-ruby-processes-what-to-do-before-you-kill-9/) is pretty useful. My process is running under docker on Ubuntu and I haven't been able to attach gdb to it unfortunately. 

I'm assuming it's our code / a gem but not ruling anything out yet.

**Comment by camallen** [comment]
Right ended up tracking my problem down . It was totally a userland loop that could not finish, **nothing to do with Puma**.

On a colleague's recommendation I added [rack-timeout](https://github.com/heroku/rack-timeout) to the app to get a stack trace via the error reporting app. Normal traffic ended up triggering the stuck state pretty quickly and i got a useful stack trace to track down the stuck loop conditions. Fixed this and haven't experienced the error for a day now ... happy days.
